05-java-mqtt/src/main/java/demo/MqttNetworkManager.java

```diff
@@ -0,0 +1,163 @@
+    private final DeviceUID deviceUID;
+    private final String address;
+    private final int port;
+    private final String clientId;
+    private final List<String> neighbors;
+    private final int qos;
+    private IMqttAsyncClient mqttClient;
+    private transient Map<DeviceUID, Map<CodePath, Object>> messages = new HashMap<>();
+
+    /**
+     * Constructor method for which uses the default config.
+     * @param deviceUID the device id.
+     * @param neighbors the node neighbors.
+     */
+    public MqttNetworkManager(final DeviceUID deviceUID, final List<String> neighbors) {
+        this(deviceUID, DEFAULT_ADDRESS, DEFAULT_PORT, neighbors);
+    }
+
+    /**
+     * Constructor method to explicitly specify the broker addess and port.
+     * @param deviceUID the device id.
+     * @param address the MQTT broker address.
+     * @param port the MQTT broker port.
+     * @param neighbors the node neighbors.
+     */
+    public MqttNetworkManager(final DeviceUID deviceUID, final String address, final int port,
+                              final List<String> neighbors) {
+        this(deviceUID, address, port, DEFAULT_QOS, neighbors);
+    }
+
+    /**
+     * Constructor method to explicitly set the MQTT qos.
+     * @param deviceUID the device id.
+     * @param address the MQTT broker address.
+     * @param port the MQTT broker port.
+     * @param qos the MQTT qos parameter.
+     * @param neighbors the node neighbors.
+     */
+    public MqttNetworkManager(final DeviceUID deviceUID, final String address, final int port,
+                              final int qos, final List<String> neighbors) {
+        this.deviceUID = deviceUID;
+        this.address = address;
+        this.port = port;
+        this.clientId = Integer.toString(deviceUID.hashCode());
+        this.qos = qos;
+        this.neighbors = new ArrayList<>(neighbors);
+    }
+
+    /**
+     * Starts the MQTT client and subscribes to the target topic.
+     * @param topic the topic the node is listening.
+     * @throws MqttException if some MQTT related error occur.
+     * @return the token to track the listen asynchronous operation.
+     */
+    public IMqttToken listen(final String topic) throws MqttException {
+        final String broker = "tcp://" + this.address + ":" + this.port;
+        final MqttClientPersistence persistence = new MemoryPersistence();
+        this.mqttClient = new MqttAsyncClient(broker, this.clientId, persistence);
+        mqttClient.connect().waitForCompletion();
+        return mqttClient.subscribe(topic, this.qos, null, null, this::handleMessage);
+    }
+
+    /**
+     * Shutdowns the MQTT client.
+     * @return the token to track the asynchronous task
+     * @throws MqttException if some MQTT related error occur
+     */
+    public IMqttToken stop() throws MqttException {
+        return this.mqttClient.disconnect();
+    }
+
+    @SuppressWarnings("PMD.UnusedFormalParameter")
```

Comment:
Why this suppression?

Reply:
Because I pass that method as a callback to the `mqttClient.subscribe` method and I don't need the first argument. Is there another way to do it?

Reply:
Yes, wrapping it in a lambda
```java
private void handleMessage(final MqttMessage message) { ... }
mqttClient.subscribe(topic, this.qos, null, null, (_, message) -> handleMessage(message));
```

(looks soooo much better in Kotlin...)

Reply:
I get back "As of Java 9, '_' is a keyword, and may not be used as an identifier"
and I get "The lambda parameter name '_unused' doesn't match '[a-z][a-zA-Z0-9]*'" if i use something starting with '\_'.
For now i just left 'unused' as the identifier.